src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,462 @@
+// adapted from Immix
+
+use atomic::Ordering;
+
+use super::BlockList;
+use super::MarkSweepSpace;
+use crate::util::heap::chunk_map::*;
+use crate::util::linear_scan::Region;
+use crate::vm::ObjectModel;
+use crate::{
+    util::{metadata::side_metadata::SideMetadataSpec, Address, OpaquePointer, VMThread},
+    vm::VMBinding,
+};
+
+use std::num::NonZeroUsize;
+
+/// A 64KB region for MiMalloc.
+/// This is also known as MiMalloc page. We try to avoid getting confused with the OS 4K page. So we call it block.
+/// This is a non-zero block, and `Option<Block>` can still a word size value.
+// TODO: What if we actually use the first block?
```

Comment:
I think it is unlikely. But just in case, we may have to use `Block(Address)`. Then we can distinguish `Block(Address::ZERO)` and `None`. I changed the comments a little bit.

Reply:
I also have some thoughts of introducing `NonZeroAddress` which wraps `NonZeroUsize` and similar things for `Chunk` and other things.  That's for another refactoring, and I'll write it up in an issue soon.

Reply:
I thought about that when I refactored the code based on your opinion. It may not be a good idea though. In the Rust Immix paper, we wrote this:
```quote
A safer alternative in Rust is to use Option<Address> initialized as None
to indicate that there is no valid value. However, this adds an additional conditional
and a few run-time checks to extract the actual address value in the
performance-critical path of allocation, which adds around 4% performance overhead.
We deem this tradeoff not to be worthwhile given the paramount importance of the allocation fast path...
```

Reply:
The unsigned and signed NonZeroXxxx types were only introduced in 1.28.0 and 1.34.0, respectively, in the year 2018 and 2019, respectively.  I think that explains why there was a performance penalty when that paper was published in 2016, because the compiler was not able to optimise for `Option<X>` where X is just an integer.

Reply:
By the way, the penalty may be worse if a function returns `Option<Address>` because that, according to the calling convention, requires the value to be returned on the stack (or in two registers on some architectures), and that could be a real performance hit.

Reply:
That could be one part of the reason. Another big reason is that for bump pointer allocator, we initialize `cursor` and `limit` to `Address::ZERO`. We could simply check if `cursor + size < limit` (which works for both initial values and valid values. If we use `Option<Address>` (it doesn't matter `Address` is `usize` or `NonZeroUsize`), we will need an extra check to see if `cursor/limit` is `Some`.